src/scenic/core/regions.py

```diff
@@ -281,883 +306,2816 @@ def __hash__(self):
+            # If the candidate points are farther apart than the sum of the circumradius
+            # values, they can't intersect.
+
+            # Get a candidate point from each mesh. If the center of the object is in the mesh use that.
+            # Otherwise try to sample a point as a candidate, skipping this pass if the sample fails.
+            if self.containsPoint(Vector(*self.mesh.bounding_box.center_mass)):
+                s_candidate_point = Vector(*self.mesh.bounding_box.center_mass)
+            elif len(samples:=trimesh.sample.volume_mesh(self.mesh, self.num_samples)) > 0:
+                s_candidate_point = Vector(*samples[0])
+            else:
+                s_candidate_point = None
+
+            if other.containsPoint(Vector(*other.mesh.bounding_box.center_mass)):
+                o_candidate_point = Vector(*other.mesh.bounding_box.center_mass)
+            elif len(samples:=trimesh.sample.volume_mesh(other.mesh, other.num_samples)) > 0:
+                o_candidate_point = Vector(*samples[0])
+            else:
+                o_candidate_point = None
+
+            if s_candidate_point is not None and o_candidate_point is not None:
+                # Compute the inradius of each object from its candidate point.
+                s_inradius = abs(trimesh.proximity.ProximityQuery(self.mesh).signed_distance([s_candidate_point])[0])
+                o_inradius = abs(trimesh.proximity.ProximityQuery(other.mesh).signed_distance([o_candidate_point])[0])
+
+                # Compute the circumradius of each object from its candidate point.
+                s_circumradius = numpy.max(numpy.linalg.norm(self.mesh.vertices - s_candidate_point, axis=1))
+                o_circumradius = numpy.max(numpy.linalg.norm(other.mesh.vertices - o_candidate_point, axis=1))
+
+                # Get the distance between the two points and check for mandatory or impossible collision.
+                point_distance = s_candidate_point.distanceTo(o_candidate_point)
+
+                if point_distance < s_inradius + o_inradius:
+                    return True
+
+                if point_distance > s_circumradius + o_circumradius:
+                    return False
+
+            # PASS 3
+            # Use Trimesh's collision manager to check for intersection.
+            # If the surfaces collide, that implies a collision of the volumes.
+            # Cheaper than computing volumes immediately.
+            collision_manager = trimesh.collision.CollisionManager()
+
+            collision_manager.add_object("SelfRegion", self.mesh)
+            collision_manager.add_object("OtherRegion", other.mesh)
+
+            surface_collision = collision_manager.in_collision_internal()
+
+            if surface_collision:
+                return True
+
+            if self.mesh.is_convex and other.mesh.is_convex:
+                # Collision manager covers all cases for convex shapes,
```

Comment:
Let's be more explicit that for convex shapes, the manager detects containment as well as surface intersections.